pulsar-common/src/main/java/com/yahoo/pulsar/common/util/collections/GrowablePriorityLongPairQueue.java

```diff
@@ -0,0 +1,423 @@
+                remove(item.first, item.second);
+                removedValues++;
+            }
+
+        }
+        return removedValues;
+    }
+
+    /**
+     * It removes all occurrence of given pair from the queue.
+     * 
+     * @param item1
+     * @param item2
+     * @return
+     */
+    public boolean remove(long item1, long item2) {
+        tailLock.lock();
+        boolean removed = false;
+        try {
+            int size = SIZE_UPDATER.get(this);
+            int index = 0;
+            for (int i = 0; i < size; i++) {
+                if (data[index] == item1 && data[index + 1] == item2) {
+                    removeAtWithoutLock(index);
+                    removed = true;
+                }
+                index = index + 2;
+            }
+        } finally {
+            tailLock.unlock();
+        }
+        return removed;
+    }
+
+    /**
+     * Removes min element from the heap
+     * 
+     * @return
+     */
+    public LongPair remove() {
+        return removeAt(0);
+    }
+
+    private LongPair removeAt(int index) {
+        tailLock.lock();
+        try {
+            return removeAtWithoutLock(index);
+        } finally {
+            tailLock.unlock();
+        }
+    }
+
+    /**
+     * it is not a thread-safe method and it should be called before acquiring a lock by a caller.
+     * 
+     * @param index
+     * @return
+     */
+    private LongPair removeAtWithoutLock(int index) {
+        if (SIZE_UPDATER.get(this) > 0) {
+            LongPair item = new LongPair(data[index], data[index + 1]);
+            data[index] = EmptyItem;
+            data[index + 1] = EmptyItem;
+            SIZE_UPDATER.decrementAndGet(this);
+            int lastIndex = SIZE_UPDATER.get(this) << 1;
+            swap(index, lastIndex);
+            minHeapify(index, lastIndex - 2);
+            return item;
+        } else {
+            return null;
+        }
+    }
+
+    public LongPair peek() {
+        tailLock.lock();
+        try {
+            if (SIZE_UPDATER.get(this) > 0) {
+                return new LongPair(data[0], data[1]);
+            } else {
+                return null;
+            }
+        } finally {
+            tailLock.unlock();
+        }
+    }
+
+    public boolean isEmpty() {
+        return SIZE_UPDATER.get(this) == 0;
+    }
+
+    public int capacity() {
+        return this.capacity;
+    }
+
+    public void clear() {
+        tailLock.lock();
+        try {
+            int size = SIZE_UPDATER.get(this);
+
+            int index = 0;
+            for (int i = 0; i < size; i++) {
+                data[index] = -1;
+                data[index + 1] = -1;
+                index = index + 2;
+            }
+
+            SIZE_UPDATER.set(this, 0);
+        } finally {
+            tailLock.unlock();
+        }
+    }
+
+    public int size() {
+        return SIZE_UPDATER.get(this);
+    }
+
+    private boolean checkAndRemoveDuplicate(int loc, int lastIndex) {
```

Comment:
Also since this function calls removeAtWithoutLock which has the precondition that caller has taken the lock, we should name the method checkAndRemoveDuplicateWithoutLock and have the same pre condition here.